sympy/geometry/ellipse.py

```diff
@@ -695,27 +692,64 @@ def is_tangent(self, o):
         True
 
         """
-        inter = None
+        if isinstance(o, Point2D):
+            return o in self
         if isinstance(o, Ellipse):
-            inter = self.intersection(o)
-            if isinstance(inter, Ellipse):
+            if o in self:
+                return False
+            elif len(self.intersection(o)) == 1:
+                return True
+            elif len(self.intersection(o)) == 2 and \
+                    ((self.bounds[0] == o.bounds[0] and self.bounds[2] == self.bounds[2]) or
+                         (self.bounds[1] == o.bounds[1] and self.bounds[3] == self.bounds[3])):
+                return True
+            else:
+                return False
+        elif isinstance(o, Line2D):
+            if len(self.intersection(o)) == 1:
+                return True
+            else:
                 return False
-            return (inter is not None and len(inter) == 1
-                    and isinstance(inter[0], Point))
-        elif isinstance(o, LinearEntity):
-            inter = self.intersection(o)
-            if inter is not None and len(inter) == 1:
-                return inter[0] in o
+        elif isinstance(o, Ray2D):
+            if len(self.intersection(o)) == 1 and not any([self.encloses_point(point) for point in o.points]):
```

Comment:
Consider case where a Ray emanates from the border of the ellipse: they share a point but they don't have the property that "the slope of the ellipse and the slope of the continuous border of the object at the point(s) of intersection are the same". This helped me see why the tangency of a Point should not be True just because it is on the boundary of the ellipse -- it doesn't have a defined curve and a corresponding slope at the point of tangency (itself).

e.g. I think the following would fail:

```
e1 = Ellipse((0,0),3,5)
r = Ray((3,0),(3,4))
e1.is_tangent(r) is False
```